system/security/digisign/digisign.cpp

```diff
@@ -0,0 +1,224 @@
+            throw MakeStringException(-1, "digiSign:EVP_DigestSignInit: %s", buff);
+        }
+
+        //add string to the context
+        if (EVP_DigestSignUpdate(RSASignCtx, (size_t*)text, strlen(text)) <= 0)
+        {
+            char buff[120];
+            ERR_error_string(ERR_get_error(), buff);
+            throw MakeStringException(-1, "digiSign:EVP_DigestSignUpdate: %s", buff);
+        }
+
+        //compute length of signature
+        size_t encMsgLen;
+        if (EVP_DigestSignFinal(RSASignCtx, nullptr, &encMsgLen) <= 0)
+        {
+            char buff[120];
+            ERR_error_string(ERR_get_error(), buff);
+            throw MakeStringException(-1, "digiSign:EVP_DigestSignFinal1: %s", buff);
+        }
+
+        //compute signature (signed digest)
+        unsigned char * encMsg = (unsigned char*) malloc(encMsgLen);
+        if (EVP_DigestSignFinal(RSASignCtx, encMsg, &encMsgLen) <= 0)
+        {
+            char buff[120];
+            ERR_error_string(ERR_get_error(), buff);
+            throw MakeStringException(-1, "digiSign:EVP_DigestSignFinal2: %s", buff);
+        }
+
+        //convert to base64
+        JBASE64_Encode(encMsg, encMsgLen, b64Signature, false);
+
+        //cleanup
+        free(encMsg);
+        EVP_PKEY_free(priKey);
+        EVP_MD_CTX_destroy(RSASignCtx);
+        return true;//success
+#else
+        throw MakeStringException(-1, "digiSign:Platform built without ZLIB");
+        return false;
+#endif
+    }
+
+
+    //Verify the given text was used to create the given digital signature
+    bool digiVerify(const char * text, const char * b64Signature)
+    {
+        if (!verifyingConfigured)
+            throw MakeStringException(-1, "digiVerify:Verifying Digital Signatures not configured");
+
+        CriticalBlock b(digiVerifyCrit);
+
+        StringBuffer publicKeyBuff;
+        publicKeyBuff.loadFile(publicKeyFile.str());
+        if (publicKeyBuff.isEmpty())
+            throw MakeStringException(-1, "digiSign:Cannot load public key file");
+
+#ifdef _USE_ZLIB
+        BIO * keybio = BIO_new_mem_buf((void*) publicKeyBuff.str(), -1);
+        if (keybio == nullptr)
+        {
+            char buff[120];
+            ERR_error_string(ERR_get_error(), buff);
+            throw MakeStringException(-1, "digiVerify:BIO_new_mem_buf: %s", buff);
+        }
+        RSA * rsa = PEM_read_bio_RSA_PUBKEY(keybio, nullptr, nullptr, nullptr);
+
+        EVP_PKEY* pubKey = EVP_PKEY_new();
+        EVP_PKEY_assign_RSA(pubKey, rsa);
+        EVP_MD_CTX * RSAVerifyCtx = EVP_MD_CTX_create();//allocate, initializes and return a digest context
+        if (EVP_DigestVerifyInit(RSAVerifyCtx, nullptr, EVP_sha256(), nullptr, pubKey) <= 0)
+        {
+            char buff[120];
+            ERR_error_string(ERR_get_error(), buff);
+            throw MakeStringException(-1, "digiVerify:EVP_DigestVerifyInit: %s", buff);
+        }
+        BIO_free_all(keybio);
```

Comment:
Is it ok to free bio here ?  Just asking because RSAVerifyCtx is used below.

Reply:
From what I understand the keybio is only needed to read the key. It tested successfully so I plan on leaving it like this. 
FYI this is what I used as a template https://eclipsesource.com/blogs/2016/09/07/tutorial-code-signing-and-verification-with-openssl/